### DISCUSSION TITLE
AOT Support?

### DISCUSSION THREAD

**Discussion by smourier** [microsoft/CsWin32#1169] [General]
Hi,

It's not clear to me what's the status of CsWin32 (in every aspect but I'm mostly interested in generated COM interfaces) vs AOT (I'm talking .NET8 and beyond https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus%2Cwindows).

Is it supported (it doesn't seem it is but ...)?
If not, will it ever be? Any timeframe/roadmap?

Thanks!

**Comment by AArnott** [comment]
Yes, we expect that AOT works with CsWin32-generated code. In fact WinForms in .NET 8 uses CsWin32 heavily.

But for AOT support, your project should include a `NativeMethods.json` file that sets `"allowMarshaling": false`. That will switch to the ComWrappers system and is trimmable/AOT friendly.

**Comment by smourier** [comment]
If I create this project (.NET 8+AOT):

```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net8.0</TargetFramework>
		<Nullable>enable</Nullable>
		<PublishAot>true</PublishAot>
	</PropertyGroup>

	<ItemGroup>
	  <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.3.49-beta">
	    <PrivateAssets>all</PrivateAssets>
	  </PackageReference>
	</ItemGroup>

</Project>

```
With a NativeMethods.txt file like this:

```
CreateDXGIFactory
IDXGIFactory

```
And a Program.cs like this:

```
using Windows.Win32;
using Windows.Win32.Graphics.Dxgi;

namespace ConsoleApp5
{
    internal class Program
    {
        static void Main()
        {
            PInvoke.CreateDXGIFactory(typeof(IDXGIFactory).GUID, out var fac).ThrowOnFailure();
        }
    }
}

```

Generation and compilation works fine, but at runtime, I get this error:

```
Unhandled exception. System.NotSupportedException: Built-in COM has been disabled via a feature switch. See https://aka.ms/dotnet-illink/com for more information.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr& ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Windows.Win32.PInvoke.CreateDXGIFactory(Guid* riid, Object& ppFactory)
   at Windows.Win32.PInvoke.CreateDXGIFactory(Guid& riid, Object& ppFactory) in E:\ConsoleApp5\Microsoft.Windows.CsWin32\Microsoft.Windows.CsWin32.SourceGenerator\Windows.Win32.PInvoke.dxgi.dll.g.cs:line 32
   at ConsoleApp5.Program.Main() in E:\ConsoleApp5\Program.cs:line 10

```

And I don't see how to fix this since the generated code doesn't use ComWrappers, or am I mistaken?

**Reply by AArnott** [reply]
Add to your repro a `NativeMethods.json` file that sets `"allowMarshaling": false`. That will switch to the ComWrappers system and is trimmable/AOT friendly.

**Comment by smourier** [comment]
Ok thanks, it wasn't obvious looking at available documentation. For reference, I can have it work with this .csproj

```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net8.0</TargetFramework>
		<Nullable>enable</Nullable>
		<PublishAot>true</PublishAot>
		<AllowUnsafeBlocks>true</AllowUnsafeBlocks>
	</PropertyGroup>

	<ItemGroup>
	  <None Remove="NativeMethods.json" />
	</ItemGroup>

	<ItemGroup>
	  <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.3.49-beta">
	    <PrivateAssets>all</PrivateAssets>
	  </PackageReference>
	</ItemGroup>

</Project>

```

NativeMethods.json:
```
{
  "$schema": "https://aka.ms/CsWin32.schema.json",
  "allowMarshaling": false,
  "comInterop": {
    "preserveSigMethods": [ "*" ]
  }
}
```

And this code:

        static unsafe void Main()
        {
            PInvoke.CreateDXGIFactory(typeof(IDXGIFactory).GUID, out var obj).ThrowOnFailure();
            var factory = (IDXGIFactory*)obj;

            IDXGIAdapter* adapter;
            var i = 0;
            while (factory->EnumAdapters((uint)i, &adapter).Succeeded)
            {
                adapter->GetDesc(out var desc);
                Console.WriteLine(desc.Description));
                i++;
            }
        }

One last question: this code has now the requirement of being "fully unsafe", is there any way to make it generate IntPtr/managed interfaces instead of raw pointers? Something like this:

        static void Main() // not unsafe (csproj still has AllowUnsafeBlocks true)
        {
            PInvoke.CreateDXGIFactory(typeof(IDXGIFactory).GUID, out IntPtr unk).ThrowOnFailure();

            var cs = new StrategyBasedComWrappers();
            var factory = (IFactory)cs.GetOrCreateObjectForComInstance(unk, CreateObjectFlags.None);

            var i = 0;
            while (factory.EnumAdapters((uint)i, out var adapter).Succeeded)
            {
                adapter.GetDesc(out var desc);
                Console.WriteLine(desc.Description);
                i++;
            }
        }

**Reply by AArnott** [reply]
(Do you know you can reply to an existing thread instead of creating a new thread with each comment you make?)

Why did you turn on `PreserveSigMethods: *`? That will generally make the APIs harder to work with.
As for returning managed interfaces instead of raw pointers, I imagine that's a fallout of turning marshaling off. It may be that in their use of CsWin32, WinForms accepts this and uses ComWrappers to do all the wrapping of pointers, but I'm not sure. @jeremykuhne is your experience similar to @smourier's, or is there something better that can already be done with CsWin32 as-is?

**Reply by smourier** [reply]
(sorry for the thread thing). I'm using PreserveSigMethods because (generally) I'm used to HRESULT-type error handling. In many COM scenarios, I find handling .NET exceptions quite overkill (not even speaking of lost cycles to handle "likely to happen" errors like in my sample code above). Plus we sometimes need to handle success HRESULTs (>0) cases like S_FALSE or values like S_CONFLICT_RESOLVED say.

As for the managed interfaces vs raw pointers, I'd like to use code like what's generated with the new "Source generation for ComWrappers" .NET 8 feature https://learn.microsoft.com/en-us/dotnet/standard/native-interop/comwrappers-source-generation Usage I find it more ".NET friendly" and less-unsafe to use it, with managed COM interfaces (and sometimes IntPtr that's ok) instead of raw * pointers.

**Reply by JeremyKuhne** [reply]
You can provide managed wrappers for native pointers via ComWrappers, but WinForms isn't doing this except to create uncastable `object` wrappers for a few public API where we expose `object` when built-in COM is disabled.

CsWin32 might be able to generate managed ComWrappers compatible wrappers (IDynamicInterfaceCastable), but that would be non-trivial to write if you want it to behave something like built-in COM. You have to know up front every interface your wrapper class should support and provide various hooks to both make the request to whatever ComWrappers instance is being used and a call back to actually provide the generated class.

Ideally we'd be able to have CsWin32 generate code that the runtime COM source generators could use, but you cannot have source generators that depend on other source generators. :/

@smourier you might want to look at what I'm doing in my [thirtytwo](https://github.com/JeremyKuhne/thirtytwo/blob/main/src/thirtytwo/Win32/Graphics/Direct2D/RenderTarget.cs) project to lightly wrap Direct2d objects in an AOT friendly manner. Might not be enough for you, but the pattern might still be interesting. Note that DirectX does not follow all COM rules and the currently generated code does not work right with returning structs (neither did built-in COM). I've hacked around some of the problematic APIs in my project.

**Reply by smourier** [reply]
@JeremyKuhne thanks, I remember I already looked at WInterop some years ago. I know DirectX and Direct2D quite well, I’ve also worked on similar stuff (just because Microsoft doesn’t provide something like TerraFx & Winterop in an official way :-):

I have [DirectN](https://github.com/smourier/DirectN) that's like TerraFx. It's more "wrappers" than a full object model but doesn't require unsafe at all and doesn't require .NET Core (supports .NET framework).
And then I have [Wice](https://github.com/aelyo-softworks/Wice) over DirectN that is a UI engine based on Direct Composition and Direct2D (no Winforms, no WPF, no WinUI3, just Win32 and the DWM).

I'm now thinking about providing an AOT version of these projects and ideally, I'd like them to be based on Win32Metadata and CsWin32 (both projects didn't exist when I started DirectN) but I'd also like to avoid * pointers like the code I have today. Ideally, I’d also like to use the newer officially generated .NET Core 8 COM Wrappers.

**Reply by AArnott** [reply]
There has been a little talk about CsWin32 'chaining in' the COM source generator from the .NET SDK, since the compiler won't do it for us. But nothing we've followed up on. Getting specifics about where CsWin32's current code generation falls short and what additional code you'd like to see generated would be very helpful for when we get time to invest more in this area.

**Reply by smourier** [reply]
I'm currently porting my DirectN project (https://github.com/smourier/DirectN) into a .NET 8 AOT-compatible version (https://github.com/smourier/DirectNAot), using a tool of my own (https://github.com/smourier/Win32InteropBuilder) that generates code based on newer CLR interop constructs . It's based on the `win32metadata `nuget, but not on CsWin32 for the reasons I gave.

In this project (work in progress), my objectives are the following:

* 100% AOT support
* Use ComWrapper source generation for COM
* Use LibraryImport for p/invoke
* Don't use unsafe C# code (obviously allow unsafe blocks for AOT support), so don't use raw pointers variables like `int*`, so restrict all pointers to IntPtr

Few remarks:
* I use InlineArray when needed for fixed sized buffers since it's compatible with AOT
* I use PreserveSig (HRESULT) everywhere I can for reasons above
* In this specific package, I unify all namespaces into one. All types are in this namespace, there's one file for all constants declaration, one file for all functions (LibraryImport). I think it's easier for coding instead of searching for namespaces but this is arguable :-)

The current version compiles but lacks some features the current has. Here are some examples: 
COM interface https://github.com/smourier/DirectNAot/blob/main/DirectNAot/Generated/DirectN/IDXGIFactory1.cs

```
[SupportedOSPlatform("windows6.1")]
[GeneratedComInterface, Guid("770aae78-f26f-4dba-a829-253c83d1b387")]
public partial interface IDXGIFactory1 : IDXGIFactory
{
    [PreserveSig]
    [return: MarshalAs(UnmanagedType.Error)]
    HRESULT EnumAdapters1(uint Adapter, out IDXGIAdapter1 ppAdapter);
    
    [PreserveSig]
    [return: MarshalAs(UnmanagedType.U4)]
    bool IsCurrent();
}
```
Constants https://github.com/smourier/DirectNAot/blob/main/DirectNAot/Generated/DirectN/Constants.cs
Functions https://github.com/smourier/DirectNAot/blob/main/DirectNAot/Generated/DirectN/Functions.cs

Structs with fixed sized array embedded https://github.com/smourier/DirectNAot/blob/main/DirectNAot/Generated/DirectN/DXGI_OUTPUT_DESC.cs:

```
[StructLayout(LayoutKind.Sequential)]
public partial struct DXGI_OUTPUT_DESC
{
    public InlineArraySystemChar32 DeviceName;
    public FoundationRECT DesktopCoordinates;
    
    [MarshalAs(UnmanagedType.U4)]
    public bool AttachedToDesktop;
    public DXGI_MODE_ROTATION Rotation;
    public HMONITOR Monitor;
}
```
```
[InlineArray(32)]
public partial struct InlineArraySystemChar32
{
    public char Data;
    
    public override readonly string ToString() => ((ReadOnlySpan<char>)this).ToString();
}
```

BTW while building constants, I came accross the Windows.Win32.Foundation.RT_RCDATA constant which is defined as a character by win32metadata (\u000A). CsWin32 generates this code which is wrong:

	internal static partial class PInvoke
	{
		internal static readonly unsafe winmdroot.Foundation.PCWSTR RT_RCDATA = (char*)(10);
	}

**Reply by AArnott** [reply]
Thanks for sharing.

> I came accross the Windows.Win32.Foundation.RT_RCDATA constant which is defined as a character by win32metadata (\u000A). CsWin32 generates this code which is wrong:

Why is it wrong? That look like one of those constants that are special values but typed as pointers for use in APIs that consume `PWSTR` or `PCWSTR` but will specially recognize this value. You say the win32metadata defines this as a character, but it actually declares it as a PWSTR (a typed pointer) whose value is 10. I believe this is what CsWin32 is also doing. What win32metadata does isn't actually valid C# directly.

> In this specific package, I unify all namespaces into one

CsWin32 used to do this too. But we were forced to honor the namespaces in the metadata when the metadata started carrying types with the same name but in different namespaces in order to avoid collisions.

**Reply by smourier** [reply]
> Why is it wrong? 

Opps sorry, my mistake!

About unified namespace, yes, but it's still interesting when you just want a part of the whole metadata, which is in the end quite frequent.

**Reply by whiskhub** [reply]
Is there an update regarding this? It would be nice if CsWin32 could generate AOT safe code that at the same time retains the "non-unsafe" helpers and interfaces.

In my naive understanding, wouldn't it be possible for CsWin32 to generate code that has the .NET source generator attributes applied? So for example, CsWin32 could generate such a COM interface:

```cs
[GeneratedComInterface]
[Guid("3faca0d2-e7f1-4e9c-82a6-404fd6e0aab8")]
internal partial interface IFoo
{
    void Method(int i);
}
```

or such a function:

```cs
[LibraryImport("user32.dll")]
internal static partial bool IsIconic(HWND hwnd);
```

You said something about "the compiler won't do source generator chaining for us" - does it mean that source generated code cannot be source for another source generator?
Again, I'm no expert here - I just envision the same developers experience currently available for non-AOT in AOT apps as well.

**Reply by AArnott** [reply]
> You said something about "the compiler won't do source generator chaining for us" - does it mean that source generated code cannot be source for another source generator?

That's correct. So we can't simply emit `[LibraryImportAttribute]` and expect the other source generator to notice, because it won't.

Our plan at the moment is that CsWin32 will _embed_ the `LibraryImportAttribute` source generator within itself and invoke it directly to generate the code. This will allow users to retain the more friendly APIs while still avoiding the marshaler (thus being AOT-ready). That work hasn't been scheduled though.

**Comment by mungojam** [comment]
Perhaps it would have been better for me to post this here: https://github.com/microsoft/CsWin32/issues/1401. I'm struggling with getting the AOT/trim-safe mode working. I'm sure the end result is very simple but my lack of knowledge around COM is blocking me, and AI attempts to fix it aren't getting all the way to a working version either (maybe they are getting closer, impossible to tell)

**Reply by smourier** [reply]
This is because CsWin32 generates this (independently from the allowMarshaling value):

    [Guid("A9927F85-A304-4390-8B23-A75F1C668600"),ComImport()]
    internal partial class ProofOfPossessionCookieInfoManager
    {
    }

When allowMarshaling is set to true, the built-in .NET marshaling uses the `ComImport` attribute to implicitely create the COM object that has a CLSID equal to the type's Guid when you do a "new".

This magic is gone when built-in marshaling is gone (AOT cases), an object is created but it's not a COM object and it's vtable pointer is null (hence the error).

You now must create the COM object "manually". So add "CoCreateInstance" in NativeMethods.txt, and do this for example:

    PInvoke.CoCreateInstance<IProofOfPossessionCookieInfoManager>(
        typeof(ProofOfPossessionCookieInfoManager).GUID, // the CLSID
        null,
        Windows.Win32.System.Com.CLSCTX.CLSCTX_ALL,
        out var provider);

**Reply by mungojam** [reply]
@smourier amazing, thank you that has worked.

I think I've got to adjust my original clean-up code, looks like just calling `->Release()` on the provider pointer does the job.

**Comment by jevansaks** [comment]
We've introduced a new mode for cswin32 to enable AOT marshalling to work better. Please see https://microsoft.github.io/CsWin32/docs/getting-started.html#enable-cswin32runasbuildtask

**Reply by smourier** [reply]
Thanks! I’ve largely moved away from CsWin32 except for small, standalone projects or POCs. It lacks many utilities, boilerplate helpers, and shows several limitations and odd design choices when working seriously with .NET, AOT, and Windows/COM interop. Win32Metadata, however, is excellent.

**Reply by jevansaks** [reply]
I'm new to this space and making changes to adapt things to work better for .NET & AOT and COM & WinRT interop scenarios. I don't know if that will make it better for you but if you have more specific feedback, I'd be curious to know what other gaps you see. Thanks!

**Reply by smourier** [reply]
Actually, I wanted to re-test CsWin32 a bit and see the effect of the new `CsWin32RunAsBuildTask` now generating `GeneratedComInterface` but it doesn't seem it changes anything. Am I misunderstanding it?

csproj:
```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net9.0</TargetFramework>
		<Nullable>enable</Nullable>
		<PublishAot>true</PublishAot>
		<InvariantGlobalization>true</InvariantGlobalization>
		<PublishTrimmed>true</PublishTrimmed>
		<CsWin32RunAsBuildTask>true</CsWin32RunAsBuildTask>
		<DisableRuntimeMarshalling>true</DisableRuntimeMarshalling>
		<PlatformTarget>x64</PlatformTarget>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Microsoft.Windows.CsWin32" Version="0.3.236">
			<PrivateAssets>all</PrivateAssets>
			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
		</PackageReference>
	</ItemGroup>

</Project>
```

json:
```
{
  "$schema": "https://aka.ms/CsWin32.schema.json",
  "allowMarshaling": false,
  "comInterop": {
    "preserveSigMethods": [ "*" ]
  }
}
```

txt:
```
CoCreateInstance
IDispatch
Shell
```

`IDispatch` is still a struct, in generated .cs, not using `GeneratedComInterface`:

```
namespace Windows.Win32
{
	namespace System.Com
	{
		[Guid("00020400-0000-0000-C000-000000000046")]
		[global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Windows.CsWin32", "0.3.236+ecd70e0509.RR")]
		internal unsafe partial struct IDispatch
			:IComIID		
...
```

**Reply by jevansaks** [reply]
Remove the line "allowMarshaling": false (or set it to true) in the nativeMethods.json and then you should see the GeneratedComInterface and LibraryImport things being generated.

**Reply by smourier** [reply]
Ok, thanks, I misunderstood the doc here https://microsoft.github.io/CsWin32/docs/getting-started.html#disable-runtime-marshalling then.

Anyway, when I do that, this is what I get now (?):

```
		[Guid("00020400-0000-0000-C000-000000000046"),InterfaceType(ComInterfaceType.InterfaceIsIUnknown),GeneratedComInterface()]
		[global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Windows.CsWin32", "0.3.236+ecd70e0509.RR")]
		internal partial interface IDispatch
		{
			void __IDispatchPlaceholder1();

			void __IDispatchPlaceholder2();

			void __IDispatchPlaceholder3();

			void __IDispatchPlaceholder4();
		}

```

**Reply by jevansaks** [reply]
Yeah ... .NET doesn't support IDispatch with GeneratedComInterface so we produce that. Win32Metadata does have the "correct" IDispatch interface but generating it is a very heavy interface for most callers who should never go through the IDispatch path. 

Are you looking to _implement_ IDispatch for a caller that will call you through IDispatch?

CC @AaronRobinsonMSFT @jkoritzinsky

**Reply by smourier** [reply]
Both.

My position is : I want to do *everything Windows* in .NET AOT which is great, and avoid C++ :-)

I want to expose `IDispatch` objects (for example a hosted WebView2 supports it when communicating with the javascript side, like I demonstrate in this sample https://github.com/smourier/WebView2Aot/blob/main/ScriptHostObjectWebView2/Utilities/DispatchObject.cs but also for legacy apps that fancy .NET Core https://github.com/smourier/ActiveN), and I also want to use it as a COM client (CoCreateInstance etc.).

I find this confusing — the .NET runtime (and newer COM generation) side state that `IDispatch` isn’t supported any more, while on CsWin32 side, it's not even generated. This is exactly what I meant when I said "several limitations and odd design choices".

**Reply by AaronRobinsonMSFT** [reply]
> I find this confusing — the .NET runtime (and newer COM generation) side state that IDispatch isn’t supported any more

`IDispatch` is absolutely supported, but not with COM source generation. The built-in COM interop using `ComImport` and suite of attributes works in most cases. There are issues with defining new interfaces that want to be exposed as `IDispatch` and dual, but consuming existing registered COM intefaces should be fine.

> I want to do everything Windows in .NET AOT

Exposing `IDispatch` via .NET AOT is a non goal and unlikely to exist. It is incredibly heavy weight and is COM reflection. It isn't something on our road map and doesn't align with how the .NET ecosystem is evolving.

**Reply by smourier** [reply]
I know I can generate IDispatch with .NET COM source generation (I actually do it), I know you don't want to do it automatically or don't even want to hear about IDispatch, but that's not the subject here. The discussion as I raised it was more about CsWin32 not letting us - mere mortals - to simply generate a standard COM interface (it happens it's named IDispatch) that's defined in Win32Metadata, precisely since the runtime has forgotten it, ie: *why CsWin32 even cares or has a special treatment about it?* that's all I'm saying.

**Reply by jevansaks** [reply]
Oh, is the issue that we take the "good" name with a "bad" version of the interface? The reason we generate an IDispatch is because when a type in the win32metadata derives from IDispatch, we must generate those placeholder entries in the vtable and have IDispatch as a thing you can QI for. Would it be better if the cswin32 generated interface is called "IDispatchPlaceholder"?

**Reply by smourier** [reply]
I'm not sure I fully understand :-) I would just expect CsWin32 to not do anything special with IDispatch, treat it like another, so that would mean 1) generate IDispatch like any other interface with its 4 methods and 2) generate IDispatch-derived types as interfaces deriving from the IDispatch generated in 1).

**Reply by jevansaks** [reply]
Many COM interfaces in win32 are "dual" and implement IDispatch but when you ask CsWin32 to generate them, the goal is to call their structured API side, and that caller will never use the IDispatch part. And as Aaron says above, IDispatch is very heavy weight and if we were to generate the real full IDispatch interface with all its methods it would be a lot of extra code (~5000 lines of C#) for people that otherwise don't care about it.

But I just had a thought which is that if you explicitly ask for IDispatch as a thing that CsWin32 generates, that could be our signal that you want the "real" interface and we would generate it for you in all its `[GeneratedComInterface]` glory so you can implement it in AOT if you want.

**Reply by smourier** [reply]
That sounds nice. I'm just asking to be able to do things myself when .NET doesn't do it :-)

BTW I've tested other things with new `CsWin32RunAsBuildTask` and the result is much more usable with AOT.

Another small remark: if I declare `IConnectionPointContainer`, it still generates a GUID pointer :

    unsafe winmdroot.Foundation.HRESULT FindConnectionPoint(global::System.Guid* riid, out winmdroot.System.Com.IConnectionPoint ppCP);

It would be better if this was generated as an `in Guid` like it's done in most places, avoiding the need for `unsafe` C# keyword usage here.

Maybe it's due to the lack of metadata in old Windows headers that end up (or don't) in Win32Metadata. Maybe it could be an option (is there any GUID type parameter passed by value in the whole Microsoft ecosystem?).

Or ideally, it could be nice if I could customize or influence generation at method/parameter level (advanced scenario).

**Reply by jevansaks** [reply]
https://github.com/microsoft/CsWin32/pull/1526

**Reply by jevansaks** [reply]
> It would be better if this was generated as an in Guid like it's done in most places, avoiding the need for unsafe C# keyword usage here.

They exist. Because of how COM interfaces work, we can't put the friendly methods on the COM interface. The "real" COM interface has the ABI signature. Friendly methods live on an extension class. If you use intellisense you can find the other overloads:

```c#
	[global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Windows.CsWin32", "0.3.238+3a17ab07ea.D")]
	internal static partial class System_Com_IConnectionPointContainer_Extensions
	{
		/// <inheritdoc cref="winmdroot.System.Com.IConnectionPointContainer.FindConnectionPoint(global::System.Guid*, out winmdroot.System.Com.IConnectionPoint)"/>
		internal static unsafe void FindConnectionPoint(this winmdroot.System.Com.IConnectionPointContainer @this, in global::System.Guid riid, out winmdroot.System.Com.IConnectionPoint ppCP)
		{
			fixed (global::System.Guid* riidLocal = &riid)
			{
				@this.FindConnectionPoint(riidLocal, out ppCP);
			}
		}
	}
```

At some point I do want to improve these signatures so that when folks _implement_ the types they also get to use the good signature.

**Comment by catmanjan** [comment]


**Reply by catmanjan** [reply]
Nevermind, I had to add <DisableRuntimeMarshalling>true</DisableRuntimeMarshalling> to all projects using cswin32, I thought it only had to be on the main project
